### DISCUSSION TITLE
hi Pryaxis, it seems there are some shortcomings, the item stack (duplicate item) anti-cheat system only functions to detect stacks of coins, but does not function to detect duplicate items

### DISCUSSION THREAD

**Discussion by Hex4gonForc3** [Pryaxis/TShock#3000] [General]
hi Pryaxis, it seems there are some shortcomings, the item stack (duplicate item) anti-cheat system only functions to detect stacks of coins, but does not function to detect duplicate items,I made a terraria mobile server and mobile players mostly use item stack cheats (duplicate items) and god mode cheats, the most annoying thing is they often use stack item cheats

**Comment by Arthri** [comment]
Making anti-cheat is hard, especially if you're not the game's developer. While a decent anti-cheat for Terraria is possible, there is ultimately two major limiting factors. TShock is a community of open-source developers who dedicate their free time to this project, and while some of us receive donations through Patreon or other means, it is ultimately not our full-time job and it is hard to dedicate time to TShock to work on big projects. Another limiting factor is Terraria's network protocol, which is client-first and prone to breaking under strict conditions required by a decent anti-cheat. Some people have developed decent anti-cheat solutions but they're not publicly available(either through source and compiled) because people don't tend to give out expensive things for free

**Reply by Hex4gonForc3** [reply]
Uh, but it looks like the tshock team can make that anti-cheat system, because the current version already exists even though it has a few shortcomings,They just need to update and add new systems

**Reply by Arthri** [reply]
The current version aims to prevent DoS attacks from exploits in the network protocol. The last change to it is a relatively massive project which prevents almost all exploits surrounding a certain packet. A decent anti-cheat would require a massive rewrite and restructuring of the current system. A "few" shortcomings is an understatement in the larger picture.
